### PR TITLE
[#12658] Instructor Home Page: Dropdown buttons on mobile

### LINKED ISSUE #12658: Instructor Home Page: Dropdown buttons on mobile
The dropdown buttons on mobile overflow / are clipped at the edge of the screen.

Example of overflow (note that it extends beyond the full-width footer):
overflow 1

Example of clipping by the side:
overflow 2

(I'm using width 467)

The dropdown should be within the screen width itself.

### PR DESCRIPTION
**Pull request by Tejasker**
Fixes #12658 

**Outline of Solution**
**Made  Responsive Drop-down Menu for the Instructor Home Page**
### Before there was an Overflow of the drop-down buttons on the home page

Screenshot 2023-12-16 113322-instructor
Screenshot 2023-12-16 113409-sessions
### After making it Responsive according to the size of the screen, I am able to fix the overflow(spilling out) issue
**Modified Screens**
Screenshot 2023-12-16 113640
Screenshot 2023-12-16 113720

- Tested with all screen widths.
- Run all the test cases including  lint cases & passed

### CODE REVIEW

**Comment by weiquu** [comment]
Hi @Tejasker, can you check out the failing tests? You might need to check out the relevant `spec.ts` file for the failing component tests (should be `instructor-home-page.component.spec.ts`), and the `InstructorHomePageE2ETest.java` file for the failing E2E tests. You can see more details on where it failed by clicking on the "Details" button next to the failed checks.

I've taken a quick look and it seems that the tests can't find the archive and delete buttons by the `.btn-archive-course` and `.btn-delete-course` classes. I assume it's something to do with the container they are in (i.e. they are no longer a child of the component we are looking at), but I'm not too sure. Do check on this and let us know if you are facing any issues (and what you have tried).

There are also a couple of snapshot tests that are failing. You can update the snapshots by running `npm run test` and pressing `a` to run all test cases. After that, check through the snapshots to make sure the changes are as expected, then press `u` to update them. You can find more details on snapshot testing [here](https://teammates.github.io/teammates/snapshot-testing.html).

**Comment by Tejasker** [comment]
hii @weiquu, I have made updates in the snapshot by running the npm test & checked with the component test failing cases. These component tests are now passed after updating the snapshot & also E2E tests .

**Comment by weiquu** [comment]
Hi @Tejasker, thanks for fixing the snapshot tests. Note that the other tests are still failing - please refer to the first 2 paragraphs of my previous comment for some guiding info.

In particular, the `container="body"` line is causing the dropdown to be added as the last child of the `body` element. This means that previous ways to find the button for testing no longer work. You can take a look at the Instructor Feedback Sessions page for an example. Lines 243-258 of `InstructorFeedbackSessionsPage.java` should also give you an idea of how to fix the E2E tests - these should then be transferrable to the component tests.

**Comment by Tejasker** [comment]
hii @weiquu , i have manged to solve the Component test that are causing test cases to fail in CI , I will be further working on the E2E test that Is making CI fail , if possible I required your help in solving this issue if you can guide to over come the E2E test it would be much appreciated. Anyhow i will try my best to solve the E2E test failure.

Thanks

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html [hunk 1]
```diff
@@ -39,14 +39,14 @@ <h1>Home</h1>
   <div *tmIsLoading="!hasCoursesLoaded || isCopyLoading">
     <div *ngIf="courseTabModels.length > 0">
       <div id="course-tab-{{ idx }}" class="card" *ngFor="let courseTabModel of courseTabModels; let idx = index">
-        <div class="card-header bg-primary text-white cursor-pointer"
+        <div class="card-header bg-primary text-white cursor-pointer overflow-hidden"
```
Comment (weiquu) [inline src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:42]: Just checking, is this addition needed? Seems to work fine without it. (Not sure if hidden is the right behaviour anyway - auto might be better)

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html [hunk 2]
```diff
@@ -39,14 +39,14 @@ <h1>Home</h1>
...
           (click)="courseTabModel.isTabExpanded = handleClick($event, courseTabModel); this.loadFeedbackSessions(idx);">
           <b class="course-details text-break">[{{ courseTabModel.course.courseId }}]: {{
             courseTabModel.course.courseName }}</b>
           <div class="card-header-btn-toolbar flex-lg-shrink-0" *ngIf="courseTabModel.isAjaxSuccess">
-            <span ngbDropdown>
+            <span ngbDropdown container="body">
               <button type="button" class="btn btn-primary btn-sm" ngbDropdownToggle> Students </button>
-              <div ngbDropdownMenu (click)="$event.stopPropagation()">
+              <div ngbDropdownMenu  (click)="$event.stopPropagation()">
```
Comment (weiquu) [inline src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:49]: Let's remove unnecessary formatting changes - not sure why a space here is needed

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html [hunk 3]
```diff
@@ -57,23 +57,23 @@ <h1>Home</h1>
                 </a>
               </div>
             </span>
-            <span ngbDropdown>
+            <span ngbDropdown container="body">
               <button type="button" class="btn btn-primary btn-sm" ngbDropdownToggle> Instructors </button>
-              <div ngbDropdownMenu (click)="$event.stopPropagation()">
+              <div ngbDropdownMenu  (click)="$event.stopPropagation()">
```
Comment (weiquu) [inline src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:62]: Same as above - remove unnecessary formatting changes

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html [hunk 4]
```diff
@@ -57,23 +57,23 @@ <h1>Home</h1>
...
                 <a class="btn btn-light btn-sm dropdown-item" tmRouterLink="/web/instructor/courses/edit"
-                  [queryParams]="{courseid: courseTabModel.course.courseId}"> View / Edit
+                   [queryParams]="{courseid: courseTabModel.course.courseId}"> View / Edit
```
Comment (weiquu) [inline src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:64]: Same as above - remove unnecessary formatting changes

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html [hunk 5]
```diff
@@ -57,23 +57,23 @@ <h1>Home</h1>
...
                 </a>
               </div>
             </span>
-            <span ngbDropdown>
+            <span ngbDropdown container="body">
               <button type="button" class="btn btn-primary btn-sm" ngbDropdownToggle> Sessions </button>
-              <div ngbDropdownMenu (click)="$event.stopPropagation()">
+              <div ngbDropdownMenu  (click)="$event.stopPropagation()">
```
Comment (weiquu) [inline src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:70]: Same as above - remove unnecessary formatting changes

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts [hunk 6]
```diff
@@ -233,8 +238,13 @@ describe('InstructorHomePageComponent', () => {
...
     courseButton.click();
-    const archiveButton: any = fixture.debugElement.nativeElement.querySelector('.btn-delete-course');
-    archiveButton.click();
+    const deleteButton: any = document.querySelector('body > div > div > a.btn-delete-course.btn');
+      // Checking deleteButton
+      if (deleteButton) {
+        deleteButton.click();
+      } else {
+        console.error('Delete button not found');
+      }
```
Comment (weiquu) [inline src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts:247]: Let's remove debugging before merging in - just `deleteButton.click()` should be fine

File: src/e2e/java/teammates/e2e/pageobjects/InstructorHomePage.java [hunk 7]
```diff
@@ -119,14 +119,14 @@ public void deleteSession(int courseTabIndex, int sessionIndex) {
     public void archiveCourse(int courseTabIndex) {
         WebElement courseTab = getCourseTab(courseTabIndex);
         click(courseTab.findElement(By.className("btn-course")));
-        clickAndConfirm(courseTab.findElement(By.className("btn-archive-course")));
+        clickAndConfirm(browser.driver.findElement(By.cssSelector("body > div > div >a.btn-archive-course.btn")));
```
Comment (weiquu) [inline src/e2e/java/teammates/e2e/pageobjects/InstructorHomePage.java:122]: Can this be shortened to `body > div > div > .btn-archive-course`? Seems a bit more readable to me, and conveys the same meaning.

File: src/e2e/java/teammates/e2e/pageobjects/InstructorHomePage.java [hunk 8]
```diff
@@ -119,14 +119,14 @@ public void deleteSession(int courseTabIndex, int sessionIndex) {
...
         waitUntilAnimationFinish();
     }
 
     public void deleteCourse(int courseTabIndex) {
         WebElement courseTab = getCourseTab(courseTabIndex);
         click(courseTab.findElement(By.className("btn-course")));
-        clickAndConfirm(courseTab.findElement(By.className("btn-delete-course")));
+        clickAndConfirm(browser.driver.findElement(By.cssSelector("body > div > div > a.btn-delete-course.btn")));
```
Comment (weiquu) [inline src/e2e/java/teammates/e2e/pageobjects/InstructorHomePage.java:129]: Same as above - can this be shortened to `body > div > div > .btn-delete-course`? Seems a bit more readable to me, and conveys the same meaning.

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html (hunk 1 above)
Comment (Tejasker) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:42]: I thought I was appending to the body so I made it on purpose, but overflow-auto should also work as mentioned.
tested it & working as required

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html (hunk 2 above)
Comment (Tejasker) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:49]: Definitely, I will format all the similar scenarios

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts [hunk 9]
```diff
@@ -205,8 +205,13 @@ describe('InstructorHomePageComponent', () => {
...
     courseButton.click();
-    const archiveButton: any = fixture.debugElement.nativeElement.querySelector('.btn-archive-course');
-    archiveButton.click();
+    const archiveButton: any = document.querySelector('body > div > div >a.btn-archive-course.btn ');
+    // Checking archiveButton
+    if (archiveButton) {
+      archiveButton.click();
+    } else {
+      console.error('Archive button not found');
+    }
```
[the remark this answers is not in the document]
Comment (Tejasker) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts:214]: Sure i can do that

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts (hunk 6 above)
Comment (Tejasker) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts:247]: Sure I can do that

File: src/e2e/java/teammates/e2e/pageobjects/InstructorHomePage.java (hunk 7 above)
Comment (Tejasker) [reply src/e2e/java/teammates/e2e/pageobjects/InstructorHomePage.java:122]: No & I have mentioned the reason for it too below 
you can see it in [instructor-home-page.component.spec.ts] for archive button

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts [hunk 10]
```diff
@@ -233,8 +238,13 @@ describe('InstructorHomePageComponent', () => {
 
     const courseButton: any = fixture.debugElement.nativeElement.querySelector('.btn-course');
     courseButton.click();
-    const archiveButton: any = fixture.debugElement.nativeElement.querySelector('.btn-delete-course');
-    archiveButton.click();
+    const deleteButton: any = document.querySelector('body > div > div > a.btn-delete-course.btn');
```
[the remark this answers is not in the document]
Comment (Tejasker) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts:241]: @weiquu, I know this naming ie the query selector is long & more readable But the actual problem is when I try to shortened it , it doesn't seems to work as expected I have tried in multiple ways and tested , Each and every time the shortened version of query selector is failing 

### Success Case
Success-Screenshot 2023-12-19 210724
### Failure Cases
fail-Screenshot 2023-12-19 210930

Screenshot 2023-12-19 211454

Screenshot 2023-12-19 211533

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts [hunk 11]
```diff
@@ -205,8 +205,13 @@ describe('InstructorHomePageComponent', () => {
 
     const courseButton: any = fixture.debugElement.nativeElement.querySelector('.btn-course');
     courseButton.click();
-    const archiveButton: any = fixture.debugElement.nativeElement.querySelector('.btn-archive-course');
-    archiveButton.click();
+    const archiveButton: any = document.querySelector('body > div > div >a.btn-archive-course.btn ');
```
[the remark this answers is not in the document]
Comment (Tejasker) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.spec.ts:208]: @weiquu, I know this naming ie the query selector is long & more readable But the actual problem is when I try to shortened it , it doesn't seems to work as expected I have tried in multiple ways and tested , Each and every time the shortened version of query selector is failing 

### Success Case
Success-Screenshot 2023-12-19 210724
### Failure Cases
fail-Screenshot 2023-12-19 210930

Screenshot 2023-12-19 211454

Screenshot 2023-12-19 211533


**So finally I thought of keeping as it is ie the full path for selecting the element**

**Comment by Tejasker** [comment]
Hii @weiquu  once you are OK with my  above responses for the change requested, Then I can raise a PR

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html (hunk 1 above)
Comment (weiquu) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:42]: I think in generally, we would rather have the user see the overflow, since at least they can still use the functionality.

In this case, can just remove the `overflow-` entirely. The default behaviour (i.e. no change) is fine

**Comment by weiquu** [comment]
> Hii @weiquu once you are OK with my above responses for the change requested, Then I can raise a PR

Hi @Tejasker, replying to the E2E and component tests comments here:

I think my suggestion of `body > div > div > .btn-delete-course` (and similarly for `.btn-archive-course`) works because we are still directing the query down the same path to the button. I've checked that this works with a test PR which you can view [here](https://github.com/weiquu/teammates/pull/7). You can verify that all checks pass, and under the [files changed](https://github.com/weiquu/teammates/pull/7/files) section, that I've used `body > div > div > .btn-delete-course`. If I'm not wrong, querySelector needs a reference to the parent element of the element we are finding, which is why the examples you've given above do not work.

Let me know if the above makes sense. If it's okay, please go ahead and make the changes to the same branch (i.e. use this same PR)

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html (hunk 1 above)
Comment (Tejasker) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:42]: if you wish , users to view the Overflow then we can remove the overflow- completely without disturbing the original functionality and it will work fine 
So I will remove the overflow

Comment (weiquu) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.html:42]: > So I will remove the overflow

Great, thanks!

**Comment by Tejasker** [comment]
Hii @weiquu  as mentioned by you above regarding the changes, I have made the necessary changes ( ie small adjustments & alignments in the code) and merged the 2 commits ahead of the my  main repo #12671, #12667, and made my forked repo in- sync with the original . I Have raised PR with the same branch as earlier.

**Review by weiquu — APPROVED** [review]
LGTM!

**Review by domlimm — APPROVED** [review]
LGTM. Great work, and thank you for your contribution!
